### DISCUSSION TITLE
Feature request: Whitelist/Blacklist of selectors to prevent applying to every single target <i> tag.

### DISCUSSION THREAD

**Discussion by jklmnop** [FortAwesome/Font-Awesome#20956] [Feature Requests]
I am currently using Editor.md on a Drupal site and FontAwesome is overriding all of the toolbar icons with SVG because it applies to every `.fa` element on the page. This effectively renders the toolbar useless because it changes the expectations of the click events and it also tries to replace "icons" that aren't really FA icons (see image below). It would be great if I could ignore the application of SVG to Editor.md.

I would like to be able to set a blacklist FontAwesomeConfig option that is maybe an array of css selectors so I could effectively ignore any `<i />` tags within a certain container.

```javascript
FontAwesomeConfig.blacklist = [
  '.my-toolbar-container'
];

// does not apply FA or SVG transforms to children of specified containers
```

Additionally, it would be great to apply the SVG transform ONLY to specific containers. 

```javascript
FontAwesomeConfig.whitelist = [
  '.my-content-container'
];

// only applies FA or SVG transforms to children of specified containers
```

I've considered using other editors, but many of my options use font awesome or `<i />` tags in their toolbars too, so I expect there to be the same issue.

![image](https://user-images.githubusercontent.com/180323/78799784-9e812b80-7988-11ea-82ec-02a8d6ee3671.png)

Obviously, if anyone has any ideas of how to achieve what I'm trying to do without a feature request, I would be grateful to try it.

Thanks!

**Feature request checklist**
- [x] This is a single feature (i.e. not a re-write of all of Font Awesome)
- [x] The title starts with "Feature request: " and is followed by a clear feature name (Ex: `Feature request: moar cowbell`)
- [x] I have [searched for existing issues](https://github.com/FortAwesome/Font-Awesome-Pro/issues) and to the best of my knowledge this is not a duplicate

**Comment by tagliala** [comment]
Hi!

Thanks for being part of the Font Awesome Community.

Have you tried  to set `autoReplaceSvg`  to `nest`?
https://fontawesome.com/how-to-use/with-the-api/setup/configuration 

PS: have you control over `fa editormd-bold` element? Why it uses `fa` class?

**Comment by jklmnop** [comment]
> Thanks for being part of the Font Awesome Community.

Love FA! Thanks for the hard work!

> Have you tried to set `autoReplaceSvg` to `nest`?

I did try that actually but it still replaces the innerHTML of the H* buttons, unfortunately. 

> PS: have you control over `fa editormd-bold` element? Why it uses `fa` class?

Not sure about their motivations. I didn’t write Editor.md, but my guess is for a hook to set all events for buttons to work the same way. The click event looks for children of the `<a>` with `.fa`,  so I’m assuming they put it there 1. To leverage the base styles of `.fa`,  and 2. To simplify how it finds for the name attribute within the event to know what formatting option to execute.

But mostly I could see not wanting to apply it globally in the event there is a 3rd party script on the page that is using some other version of FA that’s built in. Or maybe you just want to mix and match font and SVG for some weird reason. 🤷‍♀️

Thanks so much for your reply!

**Comment by tagliala** [comment]
> I did try that actually but it still replaces the innerHTML of the H* buttons, unfortunately.

Could you please provide a reproducible test case and how you are attaching events to buttons? Have you considered idempotent events?

> I didn’t write Editor.md, but my guess is for a hook to set all events for buttons to work the same way. 

Any reference to the source code of this third-party? Does it support FA5?

**Comment by jklmnop** [comment]
> Could you please provide a reproducible test case and how you are attaching events to buttons? Have you considered idempotent events?

i'm not attaching any events in my code, it's the Editor.md wysiwyg's code that's doing it. 

I guess you could probably go to [their site](https://pandao.github.io/editor.md/en.html) and inject FA5 via dev tools console into it and see what happens with default config. It's basically the same outcome. Below is a snippet so you don't have to write it.

~~~javascript
let script = document.createElement('script'); script.src = 'https://use.fontawesome.com/releases/v5.13.0/js/all.js'; document.querySelector('head').appendChild(script);
~~~

Here's what it should look like before injecting FA5
![image](https://user-images.githubusercontent.com/180323/78821002-bf587980-79a6-11ea-9dc9-64f0055b5763.png)

It should look like this afterward and clicking the buttons will no longer work.

![image](https://user-images.githubusercontent.com/180323/78820946-a0f27e00-79a6-11ea-9934-ecd8245aec17.png)


The event handler in question: https://github.com/pandao/editor.md/blob/v1.5.0/src/editormd.js#L1238

> Any reference to the source code of this third-party? 

Github: https://github.com/pandao/editor.md/tree/v1.5.0

> Does it support FA5?

Not sure. They're including the [FA font files in the repo](https://github.com/pandao/editor.md/tree/v1.5.0/fonts) and it hasn't been updated in 5 years. They could probably use an update! 

---

I get that the Editor.md library desperately needs work to support FA5 better, but the point is that there might be some pages with older codebases who can't use FA5 because there's no way to prevent this kind of thing from happening with code they don't have control over. FA5 seems to be all or nothing. Maybe that's the point and if so, go ahead and close this feature request. I _totally_ understand! 👍 

I just thought having a blacklist/whitelist might be helpful if you just absolutely **don't** want FA5 to highjack some part of the page for any reason whatsoever. That's really what this feature request is about; having more control over FA5 and not trying to simply support some random 3rd party script that's messing everything up. 🤣

**Comment by tagliala** [comment]
@jklmnop thanks, I took a fast look and I've seen the `.fa` encoded, but I didn't figure out by reading the code what does it change and why the `a` element disappears.

Anyway, a request for blacklisting/whitelisting selector is perfectly fine... and is already here.

At the moment I'm struggling in finding the original request, I've also commented there because there is an incompatibility with a third party library (maybe material related?) that uses `fab` a prefix used also by Font Awesome.

For the moment I'm leaving this open, but I really have to find the original one (and fix the title)

**Comment by jklmnop** [comment]
> Anyway, a request for blacklisting/whitelisting selector is perfectly fine... and is already here.

@tagliala That’s really (font) awesome! Did not see that coming! Thanks so much for all your hard work and for entertaining my issue and trying to help.

**Comment by jklmnop** [comment]
For anyone else who may find this to be an issue, I was able to get around this by setting `FontAwesomeConfig.observeMutations = false;`.

This means that the SVG transforms only happen on page load and not when the DOM mutates.

Really hoping to see some progress on this so we can enjoy the SVG magic on XHR or other DOM manipulations!

**Comment by tjheffner** [comment]
> I would like to be able to set a blacklist FontAwesomeConfig option that is maybe an array of css selectors so I could effectively ignore any `<i />` tags within a certain container.

+1

It would be nice to be able to tell fontawesome to ignore trying to replace `<i />` tags inside of `<code />` elements, for example.
